src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,155 @@
+
+        # TH sends the Read Request Message to the DUT to read all attributes on a given cluster and Endpoint
+        # AttributePath = [[Endpoint = Specific Endpoint, Cluster = Specific ClusterID]]
+        # On receipt of this message, DUT should send a report data action with the attribute value to the DUT.
+
+        self.print_step(2, "Send Request Message to read all attributes on a given cluster and endpoint")
+        asserts.assert_in(node_label_attr, basic_information, "AttributeList not in attributes")
+        # [(0, Clusters.BasicInformation)]
+        read_request_2 = await self.default_controller.ReadAttribute(self.dut_node_id, node_label_attr_all_path)
+        attributes = read_request_2[0]
+        basic_information = attributes[basic_info_attr]
+        asserts.assert_in(node_label_attr, basic_information, "NodeLabel not in BasicInformation")
+        asserts.assert_in(data_version_attr, basic_information, "DataVersion not in BasicInformation")
+        # The output from this command gets many more values from basic_info_attr compared to the first test -- get attribute list, expand test if needed
+        asserts.assert_in(attribute_list_attr, basic_information, "AttributeList not in BasicInformation")
+
+        # TH sends the Read Request Message to the DUT to read an attribute from a cluster at all Endpoints
+        # AttributePath = [[Cluster = Specific ClusterID, Attribute = Specific Attribute]]
+        # On receipt of this message, DUT should send a report data action with the attribute value from all the Endpoints to the DUT.
+        self.print_step(3, "Send Request Message to read one attribute on a given cluster at all endpoints")
+        # [Clusters.BasicInformation.Attributes.NodeLabel]
+        read_request_3 = await self.default_controller.ReadAttribute(self.dut_node_id, [node_label_attr])
+        attributes = read_request_3[0]
+        asserts.assert_in(node_label_attr, basic_information, "NodeLabel not in BasicInformation")
+
+        # TH sends the Read Request Message to the DUT to read a global attribute from all clusters at that Endpoint
+        # AttributePath = [[Endpoint = Specific Endpoint, Attribute = Specific Global Attribute]]
+        # On receipt of this message, DUT should send a report data action with the attribute value from all the clusters to the DUT.
+        self.print_step(4, "Send Request Message to read one global attribute from all clusters at that endpoint")
+        # The output from this command gets many more values from unit_testing_attr compared to the some other tests
+        # [(Clusters.BasicInformation.Attributes.ClusterRevision)]
+        read_request_4 = await self.default_controller.ReadAttribute(self.dut_node_id, cluster_rev_attr_path)
+        attributes = read_request_4[0]
+        basic_information = attributes[basic_info_attr]
+        asserts.assert_true(basic_information[cluster_revision_attr], 3)
```

Comment:
When I do this, it just returns this value, which isn't very helpful:

`dict_keys([0])`

Would it make more sense to read this?

`read_request_4[0].keys()`

That way, it can at least return the top-level key.

Reply:
Answered outside of this thread